src/release-plan.ts

```diff
@@ -62,12 +62,12 @@ export async function planRelease({
   releaseSpecification: ReleaseSpecification;
   today: Date;
 }): Promise<ReleasePlan> {
-  const newReleaseName = today.toISOString().replace(/T.+$/u, '');
-  const newRootVersion = [
-    today.getUTCFullYear(),
-    today.getUTCMonth() + 1,
-    today.getUTCDate(),
-  ].join('.');
+  const newReleaseDate = formatDateAsISO(today, {
+    format: 'basic',
+    representation: 'date',
+  });
+  const newReleaseNumber = project.releaseInfo.releaseNumber + 1;
```

Comment:
Hmm, shouldn't the release number only be increased if the date stays the same?

Reply:
Currently it seems like it's impossible to set the release number to 1 for a new release. It always increases the previous release number, and the previous release number can't be 1, so the first release will have a release number of 2 and it'll go up from there.

Reply:
> Hmm, shouldn't the release number only be increased if the date stays the same?

It could. My thought is that from a programmatic perspective, instead of using the date to uniquely identify a release, we use a number that is always incremented by 1 for each new release to do so. Colloquially, we'd switch it around and generally use the date as a more human-readable identifier, but then use the release number to disambiguate multiple releases on the same day. From an implementation perspective, it simplifies things a bit because it means release branches would always be formatted the same and so would release notes.

For instance, say we have two releases that end up going out today. So the root package version would be `20220826.100.0`. When the branch is merged, the resulting GitHub release would have the following release notes:

> # August 26, 2022 (R100)
>
> ## @metamask/a 1.2.0
>
> ### Added
>
> * Add feature A (#94)
>
> ## @metamask/b 0.7.0
>
> ### Added
>
> * Add feature B (#91)
>
> ### Changed
>
> * Bump `@metamask/a` to 1.2.0 (#90)

Then we'd publish version `20220826.101.0`, and the GitHub release would have the following release notes:

> # August 26, 2022 (R101)
>
> ## @metamask/a 1.2.1
>
> ### Fixed
>
> * Fix some bug (#104)
>
> ## @metamask/b 0.7.1
>
> ### Changed
>
> * Bump `@metamask/a` to 1.2.1 (#102)

If we only bumped the middle number when the date changed, that's fine, but we'd have to work out how to refer to a release which is the only instance on a particular day and one which is not.

Going back to the example, say again we have two releases that end up going out today. We release `20220826.1.0`, but what would the GitHub release notes look like? Maybe we use the same number:

> # August 26, 2022 (1)
>
> ... release notes go here ...

and then the release notes for the second release would be:

> # August 26, 2022 (2)
>
> ... release notes go here ...

So, okay, that works fine. But multi-release dates are the exception, not the rule. So in reality, it means we'll have a lot of release notes that look like:

> # August 26, 2022 (1)
>
> ...
>
> # September 1, 2022 (1)
>
> ...
>
> # October 17, 2022 (1)
>
> ...

If we use an ever-increasing release number, we wouldn't have to add extra logic to increment the number only when the date changes, and we could generate the release notes in a way that's consistent and wouldn't look weird:

> # August 26, 2022 (R100)
>
> ...
>
> # September 1, 2022 (R101)
>
> ...
>
> # October 17, 2022 (R102)
>
> ...
>
> # October 17, 2022 (R103)
>
> ...

Reply:
> My thought is that from a programmatic perspective, instead of using the date to uniquely identify a release, we use a number that is always incremented by 1 for each new release to do so.

Quoting myself here, but I realize that I could switch things around so instead of `<date>.<release-number>.0` it could be `<release-number>.<date>.0`. That might seem a little more SemVer-y because it means that the first number would act more like the "major" number instead of the second.

Reply:
Oh OK, I thought it's only purpose was to allow releasing twice on a day if we need to.

If we're going with an ever-increasing number, which is fine, do we need the date? We could just use the release number.

Reply:
>I thought it's only purpose was to allow releasing twice on a day if we need to.
Oh, and backporting. It also allowed for that.

That's also compatible with dropping the date and just using the incrementing release number, except that we'd need to use an additional digit in the version for that, like the patch version.

Reply:
> Currently it seems like it's impossible to set the release number to 1 for a new release. It always increases the previous release number, and the previous release number can't be 1, so the first release will have a release number of 2 and it'll go up from there.

Ah. Good point. Maybe to fix this the tool could look at the Git tags and if there aren't any that it recognizes as release tags, it assumes that this is the very first release. In that case it check to make sure that the release number is 1 (or whatever we want the starting value to be) and leave it alone, otherwise it increments it. I also realize I have functional tests for the "very first release" case, but they are wrong. I'll have to fix this in another PR since #23 will introduce the use of Git tags to filter the set of packages that show up in the release spec.

Reply:
To come back to the overall goals a bit, here is what I think we want from the version:

* A number that goes up.
* No constraints on how often we can release.
* A way to release a new version that is in-between two existing release versions (i.e. a backport release), in case we need to update an older package because it's used by a project that can't take on the work of upgrading to latest yet.
* SemVer compatibility on the _syntax_ level, so that tools that assume SemVer will work somewhat-correctly. Semantically we won't be following SemVer rules, but diverging on syntax seems like inviting an unnecessary headache.

Not that we need to solve all of these problems in this PR, I just mean that's what the versioning scheme needs to provide us.

And lastly, as a nice-to-have, it'd be better if this number was easy for us to understand and talk about/read. Simple is better in case we need to say it out-loud. That's all that I see in terms of requirements.

* `<date>.<release number>.0` where the release number resets each day fits all of these goals.
  * In printing the name, and in talking about it generally, we can omit the release number unless we have multiple releases that day. So in your example, we can drop the `(1)`'s in the changelog.
  * We can increment the release number for multiple releases in a day, OR for backports. It solves both of those problems
    * I'm just now realizing that this doesn't allow for a way of backporting a release on the same day it went out (e.g. R1 is published, then R2, then I want to backport to R1). Hmm. Seems unlikely but not ideal. We could use the patch number for backports here to eliminate this edge case.
* `<date>.<release number>.0` where the release number constantly goes up also solves all of these problems, as long as we can use the patch number for backports.
* `<release number>.<backport number>.0` also solves this problem

I'm OK with any of these solutions, and they're all readable-enough for me. But the ever-increasing release number plus a date option does seem like the worse of these options. We don't need two numbers that go up. I think in practice we'd just end up referring to it by the date or the release number.

Reply:
Okay, I'm hearing two concerns: the backporting workflow and the utility of the number (or the date) in the version string.

To me, there are two goals of the root version: to provide a unique name for a release (as version strings always do) and to encode information that can be used by the `publish-release` action. Right now, when the `publish-release` action constructs a title for a new GitHub release, it doesn't need to read any action parameters or environment variables to do so, it can merely reuse the version string of the root package. My idea for the core monorepo (so far) has been to go one step further and extract the date and number from the version string, then format these two things into a more "pretty" title (e.g. translate `20220906.123.0` into "Sept. 6, 2022 (R123)").

Hopefully this provides a little context for the answers below.

---

As for the release number:

> the ever-increasing release number plus a date option does seem like the worse of these options. We don't need two numbers that go up. I think in practice we'd just end up referring to it by the date or the release number. Which... is fine, but, I'm curious what advantages you see in it over the others.

Sure, good question. I think the advantages came down to a slightly simpler implementation. It wouldn't be difficult to add logic that would decide when to bump that number, but I think we'd have to give a different name to the concept so that we can document it well, as "release number" might be a bit confusing. These are minor problems, but then also:

> In printing the name, and in talking about it generally, we can omit the release number unless we have multiple releases that day. So in your example, we can drop the `(1)`'s in the changelog.

If the "release number" resets each day, then the title for the first release for a given day would not be suffixed with "(1)", and subsequent releases for the same day _would_ be suffixed — presumably, "(2)", "(3)", etc. I believe I wanted to avoid this, so that the title would be formatted the same way regardless of whether it was the first release in a sequence or the only release in a sequence, and therefore it would be easier to talk about.

---

As for backporting:

> [...] here is what I think we want from the version [...]
>
> - A way to release a new version that is in-between two existing release versions (i.e. a backport release), in case we need to update an older package because it's used by a project that can't take on the work of upgrading to latest yet.

My thought is that the date would always represent the release date, not the date that a branch was cut. So while a backport release would functionally sit in between two other versions, the name for that release wouldn't. For example, if we had two versions `20220501.1.0` and `20220801.1.0`, and we wanted to backport a fix on top of `20220501.1.0`, the resulting version string would not be `20220501.1.1` but would still be `<current date>.<number>.0`. Again, this is so that the title of the resulting GitHub release would include the current date and not an older date.

That said, I hadn't thought about this case too much, and I can see how useful it would be to know whether a particular version represents a backport of an existing version.

---

Putting these two concerns together, it appears that the final option you proposed, `<release number>.<backport number>.0`, would probably be the best one. Would the "release number" increase with each release or not? Based on your comments I'm inclined to think it would not, so assuming this is the case:

* From a documentation/implementation perspective, I still think the first number needs a different name. Perhaps we make it simple and call the whole format `<major>.<minor>.<patch>` (where `patch` is always 0), or even `<major>.<patch>.0` (where `major` is understood to mean "a number that usually increases but may not" and `patch` is understood to mean "a number that increases with a backport release, like one might do for a hotfix release elsewhere").
* If the first number usually increases, except for backport releases, does the tool need to take a `--backport` option? Without this option, the current version would be incremented by bumping the first number, but with this option, it would be incremented by bumping only the second number.
* If we removed the date from the version string, then it wouldn't make sense to translate that version string into something human-readable when constructing the GitHub release title. Perhaps we go back to using the version string as the title as we do now?

Reply:
>it appears that the final option you proposed, <release number>.<backport number>.0, would probably be the best one. Would the "release number" increase with each release or not?

Ah, sorry, with that suggestion I meant to suggest an ever-increasing release number as you've used here. Forgot to specify that. that was meant to represent [this comment](https://github.com/MetaMask/create-release-branch/pull/22#discussion_r956488979).

Reply:
>If the first number usually increases, except for backport releases, does the tool need to take a --backport option? Without this option, the current version would be incremented by bumping the first number, but with this option, it would be incremented by bumping only the second number.

Yes, I think so. I can make an issue for this

>If we removed the date from the version string, then it wouldn't make sense to translate that version string into something human-readable when constructing the GitHub release title. Perhaps we go back to using the version string as the title as we do now?

Makes sense!

Reply:
I created a ticket for the backport support here: https://github.com/MetaMask/create-release-branch/issues/32

Reply:
Following our discussion in standup, for `<release number>.<backport number>.0` we would want only `release number` to increase for normal releases and only `backport number` to increase for backport releases.

Reply:
I've updated the version in a4f69eb...6f1274c.